codec-http/src/main/java/io/netty/handler/codec/http/CombinedHttpHeaders.java

```diff
@@ -64,6 +67,42 @@ public CharSequence escape(CharSequence value) {
     }
 
     @Override
+    public HttpHeaders add(Headers<? extends CharSequence, ? extends CharSequence, ?> headers) {
+        // Override the fast-copy mechanism used by DefaultHeaders
+        checkNotNull(headers, "headers");
+        if (headers == this) {
+            throw new IllegalArgumentException("can't add to itself.");
+        }
+        if (headers instanceof CombinedHttpHeaders) {
+            if (isEmpty()) {
+                // Can use the fast underlying copy
+                super.addImpl(headers);
+            } else {
+                // Values are already escaped so don't escape again
+                for (Map.Entry<? extends CharSequence, ? extends CharSequence> header : headers) {
+                    addEscapedValue(header.getKey(), header.getValue());
+                }
+            }
+        } else {
+            for (Map.Entry<? extends CharSequence, ? extends CharSequence> header : headers) {
+                add(header.getKey(), header.getValue());
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public HttpHeaders set(Headers<? extends CharSequence, ? extends CharSequence, ?> headers) {
+        checkNotNull(headers, "headers");
```

Comment:
I agree. We should only have to check `null` if we aren't "dereferencing" in the same context.